cli/testflinger_cli/__init__.py

```diff
@@ -328,29 +402,77 @@ def submit(self):
             self.do_poll(job_id)
 
-    def submit_job_data(self, data):
+    def submit_job_data(self, data: dict):
         """Submit data that was generated or read from a file as a test job"""
         try:
             job_id = self.client.submit_job(data)
         except client.HTTPError as exc:
             if exc.status == 400:
-                raise SystemExit(
+                sys.exit(
                     "The job you submitted contained bad data or "
                     "bad formatting, or did not specify a "
                     "job_queue."
-                ) from exc
+                )
             if exc.status == 404:
-                raise SystemExit(
+                sys.exit(
                     "Received 404 error from server. Are you "
                     "sure this is a testflinger server?"
-                ) from exc
+                )
             # This shouldn't happen, so let's get more information
-            raise SystemExit(
+            sys.exit(
                 "Unexpected error status from testflinger "
-                "server: {}".format(exc.status)
-            ) from exc
+                f"server: {exc.status}"
+            )
         return job_id
 
+    def submit_job_attachments(self, job_id: str, path: Path):
+        """Submit attachments archive for a job to the server
+
+        :param job_id:
+            ID for the test job
+        :param path:
+            The path to the attachment archive
+        """
+        # defaults for retries
+        wait = self.config.get("attachments_retry_wait") or 10
+        timeout = self.config.get("attachments_timeout") or 600
+        tries = self.config.get("attachments_tries") or 3
+
+        for _ in range(tries):
+            try:
+                self.client.post_attachment(job_id, path, timeout=timeout)
+            except requests.HTTPError as error:
+                # we can't recover from these errors, give up without retrying
+                if error.response.status_code == 400:
+                    raise AttachmentError(
+                        f"Unable to submit attachment archive for {job_id}: "
+                        f"{error.response.text}"
+                    ) from error
+                if error.response.status_code == 404:
+                    raise AttachmentError(
+                        "Received 404 error from server. Are you "
+                        "sure this is a testflinger server and "
+                        "that it supports attachments?"
+                    ) from error
+                # This shouldn't happen, so let's get more information
+                sys.exit(
+                    "Unexpected error status from testflinger server "
+                    f"({error.response.status_code}): {error.response.text}"
+                )
+            except (requests.Timeout, requests.ConnectionError):
+                # recoverable errors, try again
+                time.sleep(wait)
+                wait *= 0.3
```

Comment:
Is that intended for the wait to get shorter and shorter? Usually for retry logic like this, [Exponential backoff](https://en.wikipedia.org/wiki/Exponential_backoff) is the way to go, but this seems to be the opposite of that (try more aggressively and frequently)? Or was that supposed to be `wait *= 1.3` or some other factor >= 1.0?

Reply:
Great spot, it should have been `1.3`. Will fix.